File: 3439-find-minimum-diameter-after-merging-two-trees/find-minimum-diameter-after-merging-two-trees.cpp
class Solution {
public:
    unordered_map<int, vector<int>> getAdj(vector<vector<int>>& edges){
        unordered_map<int, vector<int>> adj;
        for (vector<int>& i: edges){
            adj[i[0]].push_back(i[1]);
            adj[i[1]].push_back(i[0]);
        }
        return adj;
    }

    vector<int> dfs(int curr, int parent, unordered_map<int, vector<int>>& adj){
        int diameter = 0;
        priority_queue<int, vector<int>, greater<int>> pq;
        pq.push(0);
        pq.push(0);
        for (int node: adj[curr]){
            if (node == parent)
                continue;
            vector<int> temp = dfs(node, curr, adj);
            pq.push(temp[1]);
            pq.pop();
            diameter = max(diameter, temp[0]);
        }
        int l = pq.top();
        pq.pop();
        int r = pq.top();
        diameter = max(diameter, l + r);
        vector<int> result;
        result.push_back(diameter);
        result.push_back(1 + (l > r ? l : r));
        return result;
    }

    int minimumDiameterAfterMerge(vector<vector<int>>& edges1, vector<vector<int>>& edges2) {
        unordered_map<int, vector<int>> adj1 = getAdj(edges1);
        unordered_map<int, vector<int>> adj2 = getAdj(edges2);

        vector<int> d1 = dfs(0, -1, adj1);
        vector<int> d2 = dfs(0, -1, adj2);

        return max(max(d1[0], d2[0]), 1 + (int)ceil(d1[0]/2.0) + (int)ceil(d2[0]/2.0));
    }
};

File: 3439-find-minimum-diameter-after-merging-two-trees/find-minimum-diameter-after-merging-two-trees.java
class Solution {
    public HashMap<Integer, List<Integer>> getAdjacent (int[][] edges){
        HashMap<Integer, List<Integer>> adj = new HashMap<Integer, List<Integer>>();
        for (int[] i: edges){
            List<Integer> temp = adj.getOrDefault(i[0], new ArrayList());
            temp.add(i[1]);
            adj.put(i[0], temp);
            temp = adj.getOrDefault(i[1], new ArrayList());
            temp.add(i[0]);
            adj.put(i[1], temp);
        }
        return adj;
    }

    public int[] dfs(int curr, int parent, HashMap<Integer, List<Integer>> adj){
        int distance = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        pq.add(0);
        pq.add(0);
        for (int node: adj.getOrDefault(curr, new ArrayList<Integer>())){
            if (node == parent)
                continue;
            int[] temp = dfs(node, curr, adj);
            distance = Math.max(distance, temp[0]);
            pq.add(temp[1]);
            pq.poll();
        }
        int l = pq.poll();
        int r = pq.poll();
        distance = Math.max(distance, l + r);
        return new int[]{distance, 1 + (l > r ? l : r)};
    }

    public int minimumDiameterAfterMerge(int[][] edges1, int[][] edges2) {
        HashMap<Integer, List<Integer>> adj1 = getAdjacent(edges1);
        HashMap<Integer, List<Integer>> adj2 = getAdjacent(edges2);

        int d1[] = dfs(0, -1, adj1);
        int d2[] = dfs(0, -1, adj2);

        return (int)Math.max(Math.max(d1[0], d2[0]), 1 + Math.ceil(d1[0] / 2.0) + Math.ceil(d2[0] / 2.0));
    }
}